File: multiplayer_server/app/engine/bot_ai.ts
import type { GameState, GameCommand } from './state.js'
import { UNIT_STATS } from './state.js'
import { hexDistance } from './hex.js'

/**
 * Simple bot AI that generates commands each tick.
 *
 * Strategy:
 * - Early game: spawn workers, harvest energy
 * - Mid game: spawn scouts to find enemy, soldiers to attack
 * - Late game: all-in attack on enemy base
 */
export function generateBotCommands(state: GameState, botSlot: number): GameCommand[] {
  const commands: GameCommand[] = []
  const player = state.players[botSlot]
  const base = state.getPlayerBase(botSlot)
  if (!base) return commands

  const myUnits = [...state.units.values()].filter((u) => u.playerSlot === botSlot)
  const workers = myUnits.filter((u) => u.unitType === 'worker')
  const soldiers = myUnits.filter((u) => u.unitType === 'soldier')
  const scouts = myUnits.filter((u) => u.unitType === 'scout')
  const idleWorkers = workers.filter((u) => u.status === 'idle')
  const idleSoldiers = soldiers.filter((u) => u.status === 'idle')
  const idleScouts = scouts.filter((u) => u.status === 'idle')

  // Find nearest resource node with energy
  const resources = [...state.resources.values()].filter((r) => r.remaining > 0)
  const nearestResource = resources.sort((a, b) => {
    const distA = hexDistance(a, base)
    const distB = hexDistance(b, base)
    return distA - distB
  })[0]

  // Find enemy buildings
  const enemyBuildings = [...state.buildings.values()].filter((b) => b.playerSlot !== botSlot)

  // Find visible enemy units
  const enemyUnits = [...state.units.values()].filter((u) => u.playerSlot !== botSlot)

  // --- Spawning logic ---

  // Phase 1: Need at least 2 workers for economy
  if (workers.length < 2 && player.energy >= UNIT_STATS.worker.cost) {
    commands.push({
      playerSlot: botSlot,
      toolName: 'spawn_unit',
      toolInput: { unit_type: 'worker' },
    })
  }
  // Phase 2: Scout to find enemy
  else if (scouts.length < 1 && player.energy >= UNIT_STATS.scout.cost) {
    commands.push({
      playerSlot: botSlot,
      toolName: 'spawn_unit',
      toolInput: { unit_type: 'scout' },
    })
  }
  // Phase 3: Build army
  else if (soldiers.length < 4 && player.energy >= UNIT_STATS.soldier.cost) {
    commands.push({
      playerSlot: botSlot,
      toolName: 'spawn_unit',
      toolInput: { unit_type: 'soldier' },
    })
  }
  // Phase 4: More workers if we have army
  else if (workers.length < 4 && soldiers.length >= 2 && player.energy >= UNIT_STATS.worker.cost) {
    commands.push({
      playerSlot: botSlot,
      toolName: 'spawn_unit',
      toolInput: { unit_type: 'worker' },
    })
  }
  // Phase 5: Keep building soldiers
  else if (player.energy >= UNIT_STATS.soldier.cost) {
    commands.push({
      playerSlot: botSlot,
      toolName: 'spawn_unit',
      toolInput: { unit_type: 'soldier' },
    })
  }

  // --- Worker orders: harvest ---

  for (const worker of idleWorkers) {
    if (nearestResource) {
      commands.push({
        playerSlot: botSlot,
        toolName: 'harvest',
        toolInput: { unit_id: worker.id, resource_id: nearestResource.id },
      })
    }
  }

  // --- Scout orders: explore toward nearest enemy ---

  // Find nearest enemy start position
  const enemyStartPositions = state.mapConfig.startPositions
    .filter((_, i) => i !== botSlot)
    .sort((a, b) => {
      const distA = hexDistance(a, base)
      const distB = hexDistance(b, base)
      return distA - distB
    })
  const enemyStartPos = enemyStartPositions[0] ?? { x: 16, y: 16 }

  // Find nearest enemy base (for soldier targeting too)
  const nearestEnemyBase = enemyBuildings
    .filter((b) => b.buildingType === 'base')
    .sort((a, b) => {
      const distA = hexDistance(a, base)
      const distB = hexDistance(b, base)
      return distA - distB
    })[0]

  for (const scout of idleScouts) {
    // If we see enemies, scout around them; otherwise head toward enemy start
    if (nearestEnemyBase) {
      const enemyBase = nearestEnemyBase
      // Patrol around enemy base
      const offsetX = Math.floor(Math.random() * 6) - 3
      const offsetY = Math.floor(Math.random() * 6) - 3
      const tx = Math.max(0, Math.min(state.mapConfig.width - 1, enemyBase.x + offsetX))
      const ty = Math.max(0, Math.min(state.mapConfig.height - 1, enemyBase.y + offsetY))
      commands.push({
        playerSlot: botSlot,
        toolName: 'move_unit',
        toolInput: { unit_id: scout.id, x: tx, y: ty },
      })
    } else {
      // Head toward enemy start position
      commands.push({
        playerSlot: botSlot,
        toolName: 'move_unit',
        toolInput: { unit_id: scout.id, x: enemyStartPos.x, y: enemyStartPos.y },
      })
    }
  }

  // --- Soldier orders: attack ---

  for (const soldier of idleSoldiers) {
    // Priority: attack nearby enemies, then enemy base
    const nearbyEnemy = enemyUnits.sort((a, b) => {
      const distA = hexDistance(a, soldier)
      const distB = hexDistance(b, soldier)
      return distA - distB
    })[0]

    if (nearbyEnemy) {
      const dist = hexDistance(nearbyEnemy, soldier)
      if (dist < 10) {
        commands.push({
          playerSlot: botSlot,
          toolName: 'attack_target',
          toolInput: { unit_id: soldier.id, target_id: nearbyEnemy.id },
        })
        continue
      }
    }

    if (nearestEnemyBase) {
      commands.push({
        playerSlot: botSlot,
        toolName: 'attack_target',
        toolInput: { unit_id: soldier.id, target_id: nearestEnemyBase.id },
      })
    } else {
      // No enemy base visible — move toward enemy start
      commands.push({
        playerSlot: botSlot,
        toolName: 'move_unit',
        toolInput: { unit_id: soldier.id, x: enemyStartPos.x, y: enemyStartPos.y },
      })
    }
  }

  return commands
}

File: game/crates/bf_engine/src/bot_ai.rs
use bf_types::*;
use crate::GameState;

/// Generate bot commands for a given autopilot player slot.
pub fn generate_bot_commands(state: &GameState, bot_slot: u8) -> Vec<Command> {
    let mut commands = Vec::new();

    let my_units: Vec<&crate::Unit> = state.units.iter().filter(|u| u.player_slot == bot_slot).collect();
    let workers: Vec<&&crate::Unit> = my_units.iter().filter(|u| u.unit_type == UnitType::Worker).collect();
    let soldiers: Vec<&&crate::Unit> = my_units.iter().filter(|u| u.unit_type == UnitType::Soldier).collect();
    let scouts: Vec<&&crate::Unit> = my_units.iter().filter(|u| u.unit_type == UnitType::Scout).collect();

    let idle_workers: Vec<&&crate::Unit> = workers.iter().filter(|u| u.status == UnitStatus::Idle).copied().collect();
    let idle_soldiers: Vec<&&crate::Unit> = soldiers.iter().filter(|u| u.status == UnitStatus::Idle).copied().collect();
    let idle_scouts: Vec<&&crate::Unit> = scouts.iter().filter(|u| u.status == UnitStatus::Idle).copied().collect();

    let base = state.get_player_base(bot_slot);
    let base_pos = base.map(|b| Position::new(b.x, b.y));

    let resources: Vec<&crate::ResourceNode> = state.resource_nodes.iter().filter(|r| r.remaining > 0).collect();
    let nearest_resource = base_pos.and_then(|bp| {
        resources.iter().min_by_key(|r| hex_distance(bp, Position::new(r.x, r.y)))
    });

    let enemy_buildings: Vec<&crate::Building> = state.buildings.iter()
        .filter(|b| b.player_slot != bot_slot).collect();
    let enemy_units: Vec<&crate::Unit> = state.units.iter()
        .filter(|u| u.player_slot != bot_slot).collect();

    // Spawning priorities
    if base.is_some() {
        if workers.len() < 2 {
            commands.push(Command::SpawnUnit { player_slot: bot_slot, unit_type: UnitType::Worker });
        } else if scouts.is_empty() {
            commands.push(Command::SpawnUnit { player_slot: bot_slot, unit_type: UnitType::Scout });
        } else if soldiers.len() < 4 {
            commands.push(Command::SpawnUnit { player_slot: bot_slot, unit_type: UnitType::Soldier });
        } else if workers.len() < 4 && soldiers.len() >= 2 {
            commands.push(Command::SpawnUnit { player_slot: bot_slot, unit_type: UnitType::Worker });
        } else {
            commands.push(Command::SpawnUnit { player_slot: bot_slot, unit_type: UnitType::Soldier });
        }
    }

    // Worker orders — send idle workers to harvest
    if let Some(resource) = nearest_resource {
        for worker in &idle_workers {
            commands.push(Command::Harvest {
                player_slot: bot_slot,
                unit_id: worker.id.clone(),
                resource_id: resource.id.clone(),
            });
        }
    }

    // Scout orders — patrol toward enemy base
    let nearest_enemy_base = enemy_buildings.iter()
        .filter(|b| b.building_type == BuildingType::Base)
        .min_by_key(|b| {
            base_pos.map(|bp| hex_distance(bp, Position::new(b.x, b.y))).unwrap_or(0)
        });

    for scout in &idle_scouts {
        if let Some(enemy_base) = nearest_enemy_base {
            let tx = (enemy_base.x + 3).clamp(0, state.map_width as i32 - 1);
            let ty = (enemy_base.y + 3).clamp(0, state.map_height as i32 - 1);
            commands.push(Command::MoveUnit {
                player_slot: bot_slot,
                unit_id: scout.id.clone(),
                target_x: tx,
                target_y: ty,
            });
        } else {
            commands.push(Command::MoveUnit {
                player_slot: bot_slot,
                unit_id: scout.id.clone(),
                target_x: 16,
                target_y: 16,
            });
        }
    }

    // Soldier orders — attack nearby enemies or push toward enemy base
    for soldier in &idle_soldiers {
        let soldier_pos = Position::new(soldier.x, soldier.y);
        let nearest_enemy = enemy_units.iter()
            .min_by_key(|e| hex_distance(soldier_pos, Position::new(e.x, e.y)));

        if let Some(enemy) = nearest_enemy {
            let dist = hex_distance(soldier_pos, Position::new(enemy.x, enemy.y));
            if dist < 10 {
                commands.push(Command::AttackTarget {
                    player_slot: bot_slot,
                    unit_id: soldier.id.clone(),
                    target_id: enemy.id.clone(),
                });
                continue;
            }
        }

        if let Some(enemy_base) = nearest_enemy_base {
            commands.push(Command::AttackTarget {
                player_slot: bot_slot,
                unit_id: soldier.id.clone(),
                target_id: enemy_base.id.clone(),
            });
        } else {
            // Move toward center
            commands.push(Command::MoveUnit {
                player_slot: bot_slot,
                unit_id: soldier.id.clone(),
                target_x: 16,
                target_y: 16,
            });
        }
    }

    commands
}
